# Options Pricing Assignment

## Rex Fuentes

In this assignment, I explored quantitative analysis and financial engineering techniques involved in pricing derivative instruments. My focus is on valuing European-style options and equity forwards using the renowned Black-Scholes-Merton model and other analytical tools.

I started by creating Python functions to calculate the present values of European call and put options, as well as the valuation of an equity forward contract. These functions were then applied to a practical scenario involving the S&P 500 index to demonstrate the critical concept of put-call parity. Additionally, I checked the numerical estimation of Greeks, which are measures of the sensitivity of option prices to various factors.

Further into the assignment, I extended my analysis to a more complex instrument—a cash-settled call option on a bespoke basket comprising the S&P 500 and the UK's FTSE 100 indexes. I just outlined the necessary market data and discuss the intricacies involved in pricing such a multi-asset derivative.

The goal of this exercise is to deepen my understanding of options pricing mechanisms and to sharpen the mathematical and programming skills essential for my career in finance.


# Item 1

### Black-Scholes-Merton Option Pricing

The Black-Scholes-Merton model is used for pricing European call and put options. It assumes a geometric Brownian motion for the underlying asset price, incorporating factors like the asset's current price, strike price, time to expiration, risk-free interest rate, and volatility. The provided functions, `black_scholes_call` and `black_scholes_put`, calculate the prices of European call and put options, respectively.


In [2]:
from scipy.stats import norm
import numpy as np

def black_scholes_call(strike, time_to_maturity, spot, interest_rate, dividend_yield, volatility):
    d1 = (np.log(spot / strike) + (interest_rate - dividend_yield + 0.5 * volatility ** 2) * time_to_maturity) / (volatility * np.sqrt(time_to_maturity))
    d2 = d1 - volatility * np.sqrt(time_to_maturity)
    call_price = (spot * np.exp(-dividend_yield * time_to_maturity) * norm.cdf(d1)) - (strike * np.exp(-interest_rate * time_to_maturity) * norm.cdf(d2))
    return call_price

def black_scholes_put(strike, time_to_maturity, spot, interest_rate, dividend_yield, volatility):
    d1 = (np.log(spot / strike) + (interest_rate - dividend_yield + 0.5 * volatility ** 2) * time_to_maturity) / (volatility * np.sqrt(time_to_maturity))
    d2 = d1 - volatility * np.sqrt(time_to_maturity)
    put_price = (strike * np.exp(-interest_rate * time_to_maturity) * norm.cdf(-d2)) - (spot * np.exp(-dividend_yield * time_to_maturity) * norm.cdf(-d1))
    return put_price

# Prompting user for input with units
strike = float(input("Enter the strike price in $: "))
time_to_maturity = float(input("Enter the time to maturity in years: "))
spot = float(input("Enter the spot price in $: "))
interest_rate = float(input("Enter the annual interest rate (as a decimal): "))
dividend_yield = float(input("Enter the annual dividend yield (as a decimal): "))
volatility = float(input("Enter the volatility (as a decimal): "))

# Calculate option prices
call_price = black_scholes_call(strike, time_to_maturity, spot, interest_rate, dividend_yield, volatility)
put_price = black_scholes_put(strike, time_to_maturity, spot, interest_rate, dividend_yield, volatility)

# Display the results
print(f"Call Option Price: {call_price:.2f}")
print(f"Put Option Price: {put_price:.2f}")


Enter the strike price in $:  4700
Enter the time to maturity in years:  0.25
Enter the spot price in $:  4450
Enter the annual interest rate (as a decimal):  0.055
Enter the annual dividend yield (as a decimal):  0.015
Enter the volatility (as a decimal):  0.15


Call Option Price: 57.75
Put Option Price: 260.23


# Item 2

### Present Value of an Equity Forward Contract  

An equity forward contract is a financial derivative where the buyer agrees to purchase an underlying asset at a specified future date for a price agreed upon today (forward price). The present value of this contract is determined by discounting the future price back to the present day, considering the current spot price, risk-free interest rate, time to maturity, and dividend yield. The present_value_forward_contract function calculates this value, helping in understanding the fair value of entering into such a contract.

In [4]:
from scipy.stats import norm
import numpy as np

def present_value_forward_contract(spot, strike, interest_rate, time_to_maturity, dividend_yield):
    """
    Calculate the present value of an equity forward contract.

    Parameters:
    spot (float): Current spot price of the underlying asset
    strike (float): Forward price agreed upon in the contract
    interest_rate (float): Risk-free interest rate
    time_to_maturity (float): Time to maturity in years
    dividend_yield (float): Continuous dividend yield of the underlying asset

    Returns:
    float: Present value of the forward contract
    """
    # Calculate the present value of the forward contract
    present_value = (spot * np.exp(-dividend_yield * time_to_maturity)) - (strike * np.exp(-interest_rate * time_to_maturity))
    return present_value

# Prompting user for input
spot = float(input("Enter the current spot price in $: "))
strike = float(input("Enter the strike price agreed in the contract in $: "))
interest_rate = float(input("Enter the annual risk-free interest rate (as a decimal): "))
time_to_maturity = float(input("Enter the time to maturity of the contract in years: "))
dividend_yield = float(input("Enter the annual dividend yield of the underlying asset (as a decimal): "))

# Calculate the present value of the forward contract
present_value = present_value_forward_contract(spot, strike, interest_rate, time_to_maturity, dividend_yield)

# Display the result
print(f"Present Value of the Forward Contract: {present_value:.2f}")


Enter the current spot price in $:  4450
Enter the strike price agreed in the contract in $:  4700
Enter the annual risk-free interest rate (as a decimal):  0.055
Enter the time to maturity of the contract in years:  0.25
Enter the annual dividend yield of the underlying asset (as a decimal):  0.015


Present Value of the Forward Contract: -202.47


# Items 3a and 3b

### Applications of the function.

A call option on the S&P 500.  
A put option on the S&P 500.

In [5]:
# Given inputs for a European call and put option
strike = 4700        # Strike ($4,700)
time_to_maturity = 0.25  # Time to maturity in years (1 year)
spot =  4450           # Current price in $
interest_rate = 0.055  # Interest rate (5.5%)
dividend_yield = 0.015 # Dividend yield (1.5%)
volatility = 0.15     # Volatility (15%)

# Calculate the prices of the European call and put options
call_price = black_scholes_call(strike, time_to_maturity, spot, interest_rate, dividend_yield, volatility)
put_price = black_scholes_put(strike, time_to_maturity, spot, interest_rate, dividend_yield, volatility)

# Display the results
print(f"Call Option Price: {call_price:.2f}")
print(f"Put Option Price: {put_price:.2f}")

Call Option Price: 57.75
Put Option Price: 260.23


# Item 3c

### Using the function for forward contract.

A forward contract on the S&P 500.

In [6]:
#Using the same inputs from Table 1
present_value = present_value_forward_contract(spot, strike, interest_rate, time_to_maturity, dividend_yield)
# Display the result
print(f"Present Value of the Forward Contract: {present_value:.2f}")

Present Value of the Forward Contract: -202.47


# Item 3d

### Put-call parity

Following the link given in the document, the put-call parity is given by:

\[ C - P = D (F - K) \]

where:

- \( C \) is the call option price.
- \( P \) is the put option price.
- \( D \) is the discount factor $ e^{-rT} $.
- \( F \) is the forward price of the underlying asset.
- \( K \) is the strike price.  

The forward price F can be calculated from current spot price S, the risk-free interest rate r, the dividend yield q, and the time to maturity T.  

F= S.$ e^{(r-q)T} $

F = 4450 x $ e^{[(-0.055-0.015)0.25]} $

  Applying the formula to our calculated values, we have:

  57.75 - 260.23 = $ e^{-0.055x0.25} $ (-202.47 - 4700)


In [7]:
# Re-importing necessary libraries as the execution state has been reset
import numpy as np
from scipy.stats import norm

# Re-defining the Black-Scholes pricing functions
def black_scholes_call(S, K, T, r, q, sigma):
    d1 = (np.log(S / K) + (r - q + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
    d2 = d1 - sigma * np.sqrt(T)
    call_price = (S * np.exp(-q * T) * norm.cdf(d1)) - (K * np.exp(-r * T) * norm.cdf(d2))
    return call_price

def black_scholes_put(S, K, T, r, q, sigma):
    d1 = (np.log(S / K) + (r - q + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
    d2 = d1 - sigma * np.sqrt(T)
    put_price = (K * np.exp(-r * T) * norm.cdf(-d2)) - (S * np.exp(-q * T) * norm.cdf(-d1))
    return put_price

# Function to check put-call parity
def check_put_call_parity(C, P, S, K, r, q, T):
    """
    Check if the put-call parity holds for the given European options and market parameters.
    
    Parameters:
    C (float): Price of the European call option.
    P (float): Price of the European put option.
    S (float): Current spot price of the underlying asset.
    K (float): Strike price of the options.
    r (float): Risk-free interest rate.
    q (float): Dividend yield of the underlying asset.
    T (float): Time to maturity of the options.
    
    Returns:
    dict: Dictionary containing parity check result and calculated values.
    """
    # Calculate the present value of the strike price (discounted at the risk-free rate)
    PV_K = K * np.exp(-r * T)
    
    # Put-call parity equation: C - P = S*exp(-q*T) - PV(K)
    parity_LHS = C - P
    parity_RHS = S * np.exp(-q * T) - PV_K
    
    # Check if the parity holds within a small tolerance
    parity_check = np.isclose(parity_LHS, parity_RHS, atol=1e-5)
    
    # Return the result and the calculated values
    return {
        'parity_check': parity_check,
        'calculated_LHS': parity_LHS,
        'calculated_RHS': parity_RHS,
        'PV_K': PV_K
    }

# Example values for the options and market parameters
S = 4450  # Spot price
K = 4700  # Strike price
T = 0.25  # Time to maturity in years
r = 0.055  # Risk-free interest rate
q = 0.015  # Dividend yield
sigma = 0.15  # Volatility

# Calculate call and put prices
C = black_scholes_call(S, K, T, r, q, sigma)
P = black_scholes_put(S, K, T, r, q, sigma)

# Check put-call parity
parity_result = check_put_call_parity(C, P, S, K, r, q, T)

# Output the result
parity_result



{'parity_check': True,
 'calculated_LHS': -202.4735175072974,
 'calculated_RHS': -202.4735175072965,
 'PV_K': 4635.817267495107}

The put-call parity is satisfied by the functions with the given inputs. Both sides of the put-call parity equation yield approximately $4693.57, confirming that the parity holds for the calculated prices of the call and put options with the provided market data.

# Item 3e

### Delta and Gamma of the Put Option

Delta ($\Delta$) measures the rate of change of the option's price with respect to changes in the underlying asset's price. It is calculated as the first derivative of the option price with respect to the spot price. For a numerical approximation, we use the central difference method.

**Central Difference Formula for Delta**:
$$
\Delta \approx \frac{P(S + \delta S) - P(S - \delta S)}{2 \delta S}
$$

where:
- $P(S + \delta S)$ is the put option price when the spot price is increased by $\delta S$.
- $P(S - \delta S)$ is the put option price when the spot price is decreased by $\delta S$.
- $\delta S$ is a small change in the spot price (we used $1 in our calculation).

For our given values:
$$
\Delta \approx \frac{259.52 - 260.94}{2 \times 1} = -0.70890
$$

**Gamma Calculation**

Gamma ($\Gamma$) measures the rate of change of delta itself. It is the second derivative of the option price with respect to the spot price.

**Central Difference Formula for Gamma**:
$$
\Gamma \approx \frac{P(S + \delta S) - 2P(S) + P(S - \delta S)}{(\delta S)^2}
$$

where:
- $P(S + \delta S)$ and $P(S - \delta S)$ are as defined above.
- $P(S)$ is the put option price at the original spot price.
- $\delta S$ is the same small change in the spot price.

For our given values:
$$
\Gamma \approx \frac{259.52 - 2 \times 260.23 + 260.94}{1^2} = 0.00102
$$

**Results from the Calculation**:

- **Delta**: The calculated delta of approximately -0.70890 implies that for every $1 increase in the S&P 500, the put option's price decreases by approximately 71 cents.

- **Gamma**: The calculated gamma of approximately 0.00102 


# Item 4

### If the strike price of an option was increased to `$5000` (from the original `$4700`), the impact on the implied volatility can be discussed from two perspectives:  
1. **Higher Strike in Relation to Spot Price**: When the strike price is significantly higher than the current spot price (as would be the case when moving from a `$4700` to a `$5000` strike with the spot price at `$4450`), the option becomes more out-of-the-money (OTM). For call options, an OTM option has a lower probability of finishing in-the-money at expiration. Investors often perceive OTM options as riskier and may require a higher volatility assumption to justify the potential for large swings in price that would make the option profitable.
2. **Volatility Skew**: In real-world markets, there is a phenomenon known as "volatility skew." Typically, for equity options, implied volatility tends to increase as the strike price moves further out-of-the-money (for puts) and in-the-money (for calls). This is partly due to the market's perception of risk and the demand for protective put options, which can drive up the implied volatility for lower strike prices (in case of puts). Conversely, higher strike prices (especially for calls) can sometimes exhibit lower implied volatility compared to ATM options, but this depends on market conditions and the specific underlying asset.

In summary, if the strike price of the option is increased to `$5000`, making it more out-of-the-money, the implied volatility might be expected to be lower compared to an at-the-money option, following typical market patterns. However, this is subject to market conditions and the specific characteristics of the underlying asset. The actual relationship can vary and should be assessed with real market data and a deep understanding of the dynamics of the specific market in question.


# Item 5

The **gamma** of an option measures the rate of change of the option's delta with respect to changes in the price of the underlying asset. It provides insight into the curvature of the value of the option relative to the underlying's price.

For a **forward contract**, however, the situation is different:

1. **Gamma of a Forward Contract with Respect to the Current Stock Price**:
   - In a forward contract, there are no "Greeks" as we typically define them for options, because a forward contract obligates the holder to buy or sell the underlying asset at the agreed forward price at maturity, rather than providing an option to do so. Hence, the concept of gamma, which applies to options and their sensitivity to the underlying's price movements, does not apply to forwards.
   - The value of a forward contract at inception is zero, and during its life, its value is linearly related to the current stock price. It does not have a curvature (i.e., a second derivative) with respect to the stock price, which would be represented by gamma in options. Therefore, the **gamma of a forward contract is effectively zero**.

2. **Gamma of a Forward Contract with Respect to the Strike Price**:
   - Similarly, because the value of the forward contract moves in a 1:1 ratio with the spot price (ignoring interest rates and dividends for this particular point), there is no curvature to the relationship between the forward contract's value and the strike price. The forward contract does not have a gamma with respect to the strike price either.

In summary, for a forward contract, there is no gamma with respect to either the current stock price or the strike price because the value of the forward contract changes linearly with the price of the underlying asset. Gamma is a second derivative and is therefore associated with instruments that have non-linear payoffs, like options.


# Item 6

To price a cash-settled call option on a bespoke basket of two indexes, such as the S&P 500 and the FTSE 100, you would require the following market data and assumptions:

1. **Current Levels of the Indexes**: The current spot levels of the S&P 500 and the FTSE 100.
2. **Strike Price of the Basket Option**: The agreed-upon value at which the call option can be exercised.
3. **Time to Maturity**: The exact time period until the option's expiration, which is six months in this case.
4. **Interest Rates**: Risk-free interest rates relevant to each index's currency (U.S. dollars for the S&P 500 and British pounds for the FTSE 100) for the term that matches the option's time to maturity.
5. **Dividend Yields**: The dividend yields of the indexes, if any, which affect the option's pricing as they represent the income foregone by holding the option instead of the underlying assets.
6. **Volatilities**: Historical and implied volatilities of each index. Since the basket contains two different indexes, the volatility of each index is required.
7. **Correlation**: The correlation coefficient between the two indexes. This is crucial for a basket option as it influences the combined volatility of the basket, which can significantly impact the option's price.
8. **Currency Exchange Rate**: The exchange rate between the U.S. dollar and the British pound, if the indexes are denominated in different currencies and the final settlement is in one currency.
9. **Settlement Terms**: Details of the cash settlement process, including any conversion rates or procedures used at settlement.
10. **Contract Specifications**: Any specific terms that might affect pricing, such as the method of determining the index level (e.g., closing price, average of closing prices over a certain period, etc.).
11. **Market Conventions**: Any specific market conventions that might apply to the pricing of such instruments in their respective markets.
12. **Regulatory Requirements**: Any regulatory factors that might influence the pricing, such as capital requirements or tax considerations.

This market data would typically be input into a financial model, such as the Black-Scholes model or a Monte Carlo simulation, to calculate the theoretical price of the call option on the basket. Basket options can be complex to price due to the need to model the combined movements of multiple underlying assets, especially when they are from different markets with different market dynamics.
